Fix: follow with heading tracking mode cancels user location indicator update

GenChe opened the pull request:
### Issue Link :link:
#2134 and #3112

### Goals :soccer:
* Fix: follow with heading tracking mode cancels user location indicator update

### Implementation Details :construction:

As far as I understand, the fact that `_setDirection` doesn’t preserve the center coordinate when `UserTrackingMode` is `followWithHeading` seems to be a bug. I added the fix without trying to rework the animation flow. To me, it seems reasonable that if heading updates occur very frequently, they will interrupt the interpolated animation for location changes. However, this shouldn’t cause the blue dot to remain stuck at an old position until the heading updates stop.

### Testing Details :mag:

Tested using a slightly modified version of the sample app (View Route Simulation). I added support for heading updates along with simulated location updates.

In real-world scenarios, this should look even better, since location updates will generally be less abrupt.

https://github.com/user-attachments/assets/a0911cae-54cb-4c42-bbca-06113684d4b1

https://github.com/user-attachments/assets/7c615f9a-600d-4027-8d69-0760bb44b9b1

louwers reviewed (changes requested):
The documentation of the public `setDirection:direction:animated` API notes:

>  Changing the heading rotates the map without changing the current center
 coordinate or zoom level.

And this calls `_setDirection:direction:animated` in its implementation.

So you need to find another fix for this, because this is a breaks the `setDirection:direction:animated` method.

GenChe commented:
> The documentation of the public `setDirection:direction:animated` API notes:
> 
> > Changing the heading rotates the map without changing the current center
> > coordinate or zoom level.
> 
> And this calls `_setDirection:direction:animated` in its implementation.
> 
> So you need to find another fix for this, because this is a breaks the `setDirection:direction:animated` method.

Thanks!

I’ve updated the PR to separate the two possible use cases and apply the center coordinate only when the call is triggered by a new heading update. I also added an animation timing function, similar to [this example](https://github.com/maplibre/maplibre-native/blob/0257d53dfb5a07994a8d1db3f2d39058d47e2e57/platform/ios/src/MLNMapView.mm#L6233-L6242), as used for incremental location changes. The animation looks less choppy to me now.

I tried to dig deeper to figure out what exactly causes the user location annotation to get stuck, but I wasn’t able to pinpoint the issue. My guess is that it’s due to a desync between the `userLocation.coordinate` and the value used to calculate `userLocationAnnotationViewCenter`.

That said, I think applying the last `userLocation.coordinate` on a heading update is reasonable when `userTrackingMode` is set to `follow` or `followWithHeading` — which is the case here.

GenChe commented:
Hello @louwers 

Please tell me if you want me to change anything else in this PR.

Thanks!

louwers commented on platform/ios/src/MLNMapView.mm:
```diff
@@ -6428,7 +6445,7 @@ - (void)locationManager:(__unused id<MLNLocationManager>)manager didUpdateHeadin
             && self.userTrackingState != MLNUserTrackingStateBegan)
         {
-            [self _setDirection:headingDirection animated:YES];
+            [self _setDirection:headingDirection shouldCenter:YES animated:YES];
```
I understand `_setDirection` needs the current location so that it can do the animation in one go, but maybe consider passing the center instead of a boolean. I don't think the `_setDirection` method should have anything to do with the location manager.

GenChe replied:
I changed parameter from `shouldCenter` to `center`. And so public `setDirection` method passes `kCLLocationCoordinate2DInvalid` to private `_setDirection` to avoid centering.

GenChe replied:
@louwers is there anything else that needs to be changed or you are fine with it? 

This is a quite important feature for us, and we want to have it merged (and released) as soon as possible.

louwers replied:
Looks good now, thanks!

GenChe commented:
Hello @louwers

Thanks for enabling auto-merge, but PR should still be reviewed and accepted before merging.

GenChe commented:
thanks @louwers!

seems like pod publish step failed for release 6.13.0 and we can't get this version (because we are using CocoaPods). Can you restart pod publish step, please?
